src/picongpu/include/particles/flylite/helperFields/LocalDensityFunctors.hpp

```diff
@@ -0,0 +1,169 @@
+/* Copyright 2017 Axel Huebl
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "particles/flylite/helperFields/LocalDensity.hpp"
+#include "particles/flylite/helperFields/LocalDensity.kernel"
+
+// PMacc
+#include "pmacc_types.hpp"
+#include "static_assert.hpp"
+#include "Environment.hpp"
+#include "algorithms/ForEach.hpp"
+
+#include <string>
+#include <memory>
+
+
+namespace picongpu
+{
+namespace particles
+{
+namespace flylite
+{
+namespace helperFields
+{
+namespace detail
+{
+    /** Average a group of species to a local density
+     *
+     * Takes a single species and fills the LocalDensity with it.
+     *
+     * @tparam T_SpeciesType a picongpu::Particles class with a particle species
+     */
+    template<
+        typename T_SpeciesType
+    >
+    struct AddSingleDensity
+    {
+        using SpeciesType = T_SpeciesType;
+        using FrameType = typename SpeciesType::FrameType;
+        using ShapeType = typename GetShape< SpeciesType >::type;
+        using Density = ComputeGridValuePerFrame<
+            ShapeType,
+            particleToGrid::derivedAttributes::Density
+        >;
+
+        /** Functor
+         *
+         * @param currentStep the current time step
+         * @param fieldTmp a slot of FieldTmp to add a density to
+         */
+        void operator()(
+            uint32_t currentStep,
+            std::shared_ptr< FieldTmp > fieldTmp
```

Comment:
done, adds requirement to forward the fieldTmp in the foreach call (same in the energy histogram kernel file)